File: examples/calculator-python-hexagonal-cqrs/app/application/query/history_query_service.py
from app.application.query.history_query_repository import HistoryQueryRepository
from app.application.query.history_summary import HistorySummary
from app.domain.calculation.value import CalculationId


class HistoryQueryService:
    """Read side: only reads projections — no state changes."""

    def __init__(self, query_repository: HistoryQueryRepository) -> None:
        self._query_repository = query_repository

    def list_recent(self, limit: int = 20) -> list[HistorySummary]:
        return self._query_repository.list_recent(limit)

    def get_by_id(self, calculation_id: CalculationId) -> HistorySummary | None:
        return self._query_repository.find_by_id(calculation_id)

File: examples/calculator-java-hexagonal-cqrs/src/main/java/com/oopforge/example/calculator/application/query/HistoryQueryService.java
package com.oopforge.example.calculator.application.query;

import com.oopforge.example.calculator.domain.CalculationId;
import java.util.List;
import java.util.Optional;

/**
 * Read side: only reads projections — no state changes.
 */
public class HistoryQueryService implements CalculationHistory {

    private final HistoryQueryRepository queryRepository;

    public HistoryQueryService(HistoryQueryRepository queryRepository) {
        this.queryRepository = queryRepository;
    }

    @Override
    public List<HistorySummary> listRecent(int limit) {
        return queryRepository.listRecent(limit);
    }

    @Override
    public Optional<HistorySummary> findById(CalculationId calculationId) {
        return queryRepository.findById(calculationId);
    }
}
